Schema Error Messages

danielhelfand opened the pull request:
Adding error messages for invalid schemas and if data values violate schema.

joaopapereira committed:
add schema validation error messages

Co-authored-by Steven Locke <slocke@pivotal.io>

Signed-off-by: Steven Locke <slocke@pivotal.io>

Signed-off-by: Daniel Helfand <helfand.4@gmail.com>

Commit:
Create an error message for an empty schema but non-empty DVs

* Required a GetFile func for a Position
* Changed receiver of CheckType to pointers
* When CheckType for ArrayItem or MapItem, panic if the node cannot be
cast appropriately
* Cleaned up some TODOs

Signed-off-by: Daniel Helfand <helfand.4@gmail.com>
Signed-off-by: Cari Dean <carilynndean@gmail.com>

gcheadle-vmware commented on pkg/schema/error.go:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+type invalidError struct {
```
"invlaidError" makes me think that the error is invalid, not that the schema is invalid.

StevenLocke replied:
👍 Agreed. Renamed to invalidSchemaError

gcheadle-vmware commented on pkg/schema/error.go:
```diff
@@ -0,0 +1,158 @@
+	panic("do not use directly")
+}
+
+func (e *errorWithContent) SetContext(doc *yamlmeta.Document) error {
```
Odd how in "SetContext" what is actually happening is the content that is being set. Maybe we stick with one of either content or context?

StevenLocke replied:
Removed the function entirely :D

gcheadle-vmware commented on pkg/filepos/position.go:
```diff
@@ -39,6 +39,10 @@ func (p *Position) Line() int {
 	return *p.line
 }
 
+func (p *Position) FileString() string {
```
"FileString" felt ambiguous, maybe an alternative like FileName describes what happens in this function better?

StevenLocke replied:
Removed entirely :)
It isn't the file that matters, it's the appropriately annotated document. Went back to using AsCompactString().

gcheadle-vmware commented on pkg/schema/error.go:
```diff
@@ -0,0 +1,158 @@
+	Position *filepos.Position
+}
+
+func (e errorWithContent) Error() string {
```
When deleting this function were found that all the tests still pass and there were no compilation errors. Is the purpose of this function just to make "errorWithContent" adhere to the error interface? Why is this here?

StevenLocke replied:
Removed alongside the struct.

gcheadle-vmware commented on pkg/schema/type.go:
```diff
@@ -66,101 +72,107 @@ func (m ScalarType) GetValueType() yamlmeta.Type {
 		if !m.AllowsKey(item.Key) {
-			chk.Violations = append(chk.Violations, fmt.Sprintf("Map item '%s' at %s is not defined in schema", item.Key, item.Position.AsCompactString()))
+			chk.Violations = append(chk.Violations,
+				NewUnexpectedKeyError(item.Key.(string), item.Position, m.Position))
```
We’re casting "item.key.(string)", what makes us confident in this downcast?

gcheadle-vmware commented on pkg/cmd/template/schema_test.go:
```diff
@@ -602,7 +662,29 @@ vpc:
+
+		assertYTTWorkflowFailsWithErrorMessage(t, filesToProcess, expectedErr)
+	})
+	t.Run("null value gives hint about nullable annotation", func(t *testing.T) {
```
This test seemed identical to the one below "null given as a value", maybe this test was used to TDD the hint? Do we need both of these tests?

gcheadle-vmware commented on pkg/cmd/template/schema_test.go:
```diff
@@ -393,7 +418,31 @@ not_in_schema: "this should fail the type check!"
+
+		assertYTTWorkflowFailsWithErrorMessage(t, filesToProcess, expectedErr)
+	})
+	t.Run("multiple data values given but schema is empty", func(t *testing.T) {
```
What is the value added by this test? found it difficult to see the difference between this and the one above, "data values is given but schema is empty"

danielhelfand replied:
It mainly was added to TDD. Originally, this error contained the values file name but only one if multiple were supplied. It helps reinforce the design of the error message.

gcheadle-vmware reviewed (commented):
@jtigger and I collected notes together during our review of this PR. I have provided the in line code comments and John will provide the higher level review.

cppforlife commented on pkg/schema/error.go:
```diff
@@ -0,0 +1,158 @@
+
+func (e *errorWithContent) SetContext(doc *yamlmeta.Document) error {
+	contents, err := doc.RawDataAtLine(e.Position)
+
```
nit: remove newline

StevenLocke replied:
Removed setContext entirely.

cppforlife commented on pkg/yamlmeta/parser.go:
```diff
@@ -97,8 +97,8 @@ func (p *Parser) parseBytes(data []byte, lineCorrection int) (*DocumentSet, erro
 		}
-
 		doc := &Document{
+			RawData:  data,
```
data hold entire file afaik in this case, seems a bit weird to store file contents on a document.

StevenLocke replied:
Agreed. The "source" now lives on Position, not the Document.

cppforlife commented on pkg/workspace/template_loader.go:
```diff
@@ -187,6 +187,8 @@ func (l *TemplateLoader) EvalYAML(libraryCtx LibraryExecutionContext, file *file
 		return nil, nil, err
 	}
+
+	resultVal.(*yamlmeta.DocumentSet).SetRawBytes(docSet)
```
this feels very weird. it's also not clear from the code why this is necessary.

StevenLocke replied:
Removed by giving ownership of the original line content to Position.

cppforlife commented on pkg/schema/type.go:
```diff
@@ -264,10 +294,14 @@ func (a *ArrayType) AssignTypeTo(typeable yamlmeta.Typeable) (chk yamlmeta.TypeC
 	arrayItem, ok := typeable.(*yamlmeta.ArrayItem)
 	if !ok {
-		chk.Violations = []string{fmt.Sprintf("Expected node at %s to be a %T, but was a %T", typeable.GetPosition().AsCompactString(), &yamlmeta.ArrayItem{}, typeable)}
+		chk.Violations = append(chk.Violations,
```
stylistic nit: would be nice to keep things on one line as long as possible, and then break to second line if it really runs too long.

```
chk.Violations = append(chk.Violations, NewTypeError(
  typeToString(typeable), typeToString(a), typeable.GetPosition(), a.Position))
```

pivotaljohn replied:
Agreed. At times, splitting lines by arguments can significantly improve readability, but this doesn't strike me as such a case.

cppforlife commented on pkg/schema/type.go:
```diff
@@ -278,15 +312,18 @@ func (a ArrayItemType) AssignTypeTo(typeable yamlmeta.Typeable) (chk yamlmeta.Ty
 	} // else, at a leaf
 	return
 }
-
```
nit: add back new line

cppforlife commented on pkg/schema/type.go:
```diff
@@ -66,101 +72,107 @@ func (m ScalarType) GetValueType() yamlmeta.Type {
 	return
 }
 
+func typeToString(value interface{}) string {
```
this would be nice to invert onto yamlmeta.Typeable. e.g. Typeable could have AsDebugString() or something like this.

pivotaljohn replied:
Which "this" are you referring to?

cppforlife replied:
this function.

StevenLocke replied:
TypeWithValues, and therefore Typeable, now has a "ValueTypeAsString" method which is implemented by each of the nodes. The function is still around (though we can stop exporting it) for the "item" nodes and scalars.

cppforlife commented on pkg/cmd/template/schema_test.go:
```diff
@@ -4,13 +4,12 @@
 	cmdtpl "github.com/k14s/ytt/pkg/cmd/template"
 	"github.com/k14s/ytt/pkg/cmd/ui"
 	"github.com/k14s/ytt/pkg/files"
+	"strings"
```
nit: std lib imports separation

pivotaljohn replied:
For GoLand users:

![image](https://user-images.githubusercontent.com/12101915/106055703-9b376780-60a2-11eb-8c89-f58b41fbf54f.png)

cppforlife commented on pkg/filepos/position.go:
```diff
@@ -39,6 +39,10 @@ func (p *Position) Line() int {
 	return *p.line
 }
 
+func (p *Position) FileString() string {
```
one of the rationals for not defining FileString() or similar method was to force consumers to use AsString() or AsCompatString() to provide consistent way of referencing to file location. do we really need to point to a file and not line number?

StevenLocke replied:
I agree, it feels bad to have a one-off function for this. It's currently only used in a single error message: one where non-empty data values, and an empty schema were provided. That error looks like this:

```
data values were found in data values file(s), but schema (schema.yml) has no values defined
(hint: define matching keys from data values files(s) in the schema, or do not enable the schema feature)
```

Is there a better way to report it? It felt pretty strange to see something like

```
data values were found in data_values.yml:2, but schema.yml:1 has none defined.
(hint: define matching keys from data values files(s) in the schema, or do not enable the schema feature)
```

What lines could we point to – the first instance of a data value? The document start of a schema?

------------------

After a conversation with @pivotaljohn I was convinced that the File itself isn't meaningful, it's the Document with a given annotation which marks it as a schema. I feel good in showing AsCompactString() of the schema doc, so removed the FileString function.

cppforlife commented on pkg/schema/error.go:
```diff
@@ -0,0 +1,158 @@
+	var result string
+	position := i.Position.AsCompactString()
+	lineContent := strings.TrimSpace(i.contents)
+	colonPosition := strings.Index(lineContent, ":")
```
are we assuming that this would always be map key pair? what if key includes a ":"? is there a higher level to work with here instead of working at the string level?

pivotaljohn replied:
There are a couple of places in here that we're introducing an assumption that `MapItem.Key`'s are strings. We need to revisit and generalize.

Another noted here: https://github.com/vmware-tanzu/carvel-ytt/pull/279/files/eb67827d125a4243d43253a460b8c32eb62c17ed#r564928739

StevenLocke replied:
Errors now receive the entire node, no need to worry about whether there was a colon in the key or not. We also stopped trying to point to the value with carets so determining the colon position is no longer required.

cppforlife commented on pkg/schema/type.go:
```diff
@@ -66,101 +72,107 @@ func (m ScalarType) GetValueType() yamlmeta.Type {
-		chk.Violations = append(chk.Violations, fmt.Sprintf(violationErrorMessage, node.GetValues()[0], node))
-		return
+		// A Map must've yielded a non-MapItem which is not valid YAML
+		log.Panicf("MapItem type check was called on a non-MapItem: %#v", node)
```
should be using plain `panic(...)` for internal consistency checks.

StevenLocke replied:
Updated to use regular panic instead of log.Panic'ing to remove redundant output

cppforlife commented on pkg/yamlmeta/deep_copy.go:
```diff
@@ -3,96 +3,96 @@
 
 package yamlmeta
 
-func (n *DocumentSet) DeepCopyAsNode() Node { return n.DeepCopy() }
```
i would typically recommend isolating renaming work into separate commits from feature work. it's harder to grok if something meaningful changed.

pivotaljohn replied:
The additional effort to separate portions of a commit that are renames vs. semantic changes is real. It notably took us some time to do that with this change set.

cppforlife commented on pkg/yamlmeta/document.go:
```diff
@@ -4,27 +4,42 @@
 
-func (n *Document) AsYAMLBytes() ([]byte, error) {
-	return yaml.Marshal(convertToLowYAML(convertToGo(n.Value)))
+func (d *Document) RawDataAtLine(position *filepos.Position) (string, error) {
```
it would be good to figure out how to take advantage of existing infrastructure around correlating lines to content: https://github.com/vmware-tanzu/carvel-ytt/blob/e24944a5abee281783e1ad62a483d239a80bcc04/pkg/template/compiled_template_error.go#L183.

StevenLocke replied:
Position now has this responsibility. parse sets it up and the error messages can use it without needing to ask the Document for the content at that line.

pivotaljohn commented on pkg/yamlmeta/document_set.go:
```diff
@@ -26,32 +26,51 @@ func NewDocumentSetFromBytes(data []byte, opts DocSetOpts) (*DocumentSet, error)
 
-func (n *DocumentSet) AsBytesWithPrinter(printerFunc func(io.Writer) DocumentPrinter) ([]byte, error) {
+func (ds *DocumentSet) SetRawBytes(docSet *DocumentSet) {
+	// TODO: DocumentSet.originalBytes already contain this information
+	//       This sets RawData into the documents, but in the end it contains the full document set bits and not
+	//       a document specific raw data. Eventually it might be ok but we should consider this as it might be a problem
+	//       Some options that I see are:
+	//         rename RawData to originalBytesDocumentSet or something.
+	//         templating will split these originalBytes we have into per document string.
+	//       I think the second option would be the one I would like but I think it might be a little bit of a big change
+	//       because we will have to change the yaml parser to do this....
```
🛑  This TODO needs to be resolved before this PR can be merged.

StevenLocke replied:
Position now has the responsibility of connecting Nodes to their source. parse sets it up and the error messages can use it without needing to ask the Document for the content at that line.

pivotaljohn reviewed (dismissed):
This PR includes a mechanism to re-connect the raw lines of input with the error messages that report against the corresponding node. As written, the approach introduces too much technical debt to be merged in:

- it is too complex and disperses coupling across packages (smell: [shotgun surgery](https://en.wikipedia.org/wiki/Shotgun_surgery)). To understand how raw input that resulted in a node is reconciled with it, one must:
  1. learn that `DocumentSet` attaches a copy of the raw bytes of the document set to the document;
  2. that one must call `TypeCheck.LoadContext()` to re-attach that raw data to the error message; and
  3. that error types must invoke something like `SetContext()` which delegates to `doc.RawDataAtLine()`.
- as is noted in the TODO in `pkg/yamlmeta/document_set.go`, this approach duplicates data that's present on in the `DocumentSet` struct already.
- all other context of an error condition associated with a node comes from the node, why not the line that produced the node?

Commit:
Move source into Position as "line"

* Disambiguate from old Position.line with rename to Position.lineNum
* Allows us to remove (set/load)Context
* Removes duplicated DocSet bytes off Document
* Replace the errorWithContent struct with just a Position
* Disambiguate invalidError with rename to invalidSchemaError
* panic instead of log.Panic'ing to remove redundant output

Signed-off-by: John Ryan <jtigger@infosysengr.com>

jtigger committed:
Remove carets from schema error message

* Instead, add a title (in all caps) and description of the error
* Schema errors operate on Nodes instead of strings where applicable
* Remove schema_test from the schema package - covered in template tests
* Add PositionOfDefinition and String functions to Type interface

Signed-off-by: Steven Locke <slocke@pivotal.io>

gcheadle-vmware committed:
Print key as string in an UnexpectedKeyError

jtigger committed:
Ground invalid calls to #AssignTypeTo()

... it's misleading for readers to suggest:

- scalars are typeable
- Arrays can contain anything *except* ArrayItems
- Maps can contain anything *except* MapItems

Commit:
Improve unexpectedKeyError message
Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

gcheadle-vmware committed:
Change unexepectedKeyError message

gcheadle-vmware committed:
Update hint unexpected key error
